temporal-sdk/src/test/java/io/temporal/functional/serialization/WorkflowIdSignedPayloadsTest.java

```diff
@@ -0,0 +1,250 @@
+
+    private final SimpleActivity activityLocal =
+        Workflow.newLocalActivityStub(
+            SimpleActivity.class,
+            LocalActivityOptions.newBuilder()
+                .setStartToCloseTimeout(Duration.ofSeconds(1))
+                .build());
+
+    @Override
+    public String execute(String input) {
+      assertEquals("input", input);
+      // Side Effect
+      String sideEffectResult = Workflow.sideEffect(String.class, () -> "sideEffect");
+      assertEquals("sideEffect", sideEffectResult);
+      // Activity
+      String result = activity.execute(input);
+      assertEquals("result", result);
+      // Local Activity
+      result = activityLocal.execute(input);
+      assertEquals("result", result);
+      // Child Workflow
+      if (!Workflow.getInfo().getParentWorkflowId().isPresent()) {
+        TestWorkflows.TestWorkflow1 child =
+            Workflow.newChildWorkflowStub(TestWorkflows.TestWorkflow1.class);
+        result = child.execute(input);
+        assertEquals("result", result);
+      }
+      // continueAsNew
+      if (!Workflow.getInfo().getContinuedExecutionRunId().isPresent()) {
+        Workflow.continueAsNew(input);
+      }
+      return result;
+    }
+  }
+
+  private static class PayloadEncoderWithWorkflowIdSignature implements PayloadCodec {
+    private final ByteString METADATA_ENCODING = ByteString.copyFromUtf8("signed");
+
+    private final SerializationContext serializationContext;
+
+    public PayloadEncoderWithWorkflowIdSignature() {
+      this(null);
+    }
+
+    public PayloadEncoderWithWorkflowIdSignature(
+        @Nullable SerializationContext serializationContext) {
+      this.serializationContext = serializationContext;
+    }
+
+    @Nonnull
+    @Override
+    public List<Payload> encode(@Nonnull List<Payload> payloads) {
+      return payloads.stream().map(this::encodePayload).collect(Collectors.toList());
+    }
+
+    @Nonnull
+    @Override
+    public List<Payload> decode(@Nonnull List<Payload> payloads) {
+      return payloads.stream().map(this::decodePayload).collect(Collectors.toList());
+    }
+
+    @Nonnull
+    @Override
+    public PayloadCodec withContext(@Nonnull SerializationContext context) {
+      return new PayloadEncoderWithWorkflowIdSignature(context);
+    }
+
+    private Payload encodePayload(final Payload originalPayload) {
+      String workflowId = null;
+      assertNotNull(
+          "everything in this test should go through contextualized codecs", serializationContext);
+      if (serializationContext != null) {
+        if (serializationContext instanceof WorkflowSerializationContext) {
+          workflowId = ((WorkflowSerializationContext) serializationContext).getWorkflowId();
+        } else if (serializationContext instanceof ActivitySerializationContext) {
+          workflowId = ((ActivitySerializationContext) serializationContext).getWorkflowId();
```

Comment:
Probably worth doing something activity specific in here too just to make sure that's getting piped into the context properly.